src/test/gcp/cloudfunctions.spec.ts

```diff
@@ -259,82 +261,112 @@ describe("cloudfunctions", () => {
+              members: ["allUsers"],
+            },
+          ],
+          etag: "",
+          version: 3,
+        });
 
-      const policy = cloudfunctions.generateIamPolicy(project, serviceAccount);
+        return Promise.resolve();
+      });
 
-      expect(policy.bindings.length).to.eq(1);
-      expect(policy.bindings[0].members.length).to.eq(1);
-      expect(policy.bindings[0].members[0]).to.eq(`serviceAccount:${serviceAccount}`);
+      await expect(cloudfunctions.setInvoker("project", "function", ["public"])).to.not.be.rejected;
+      expect(apiRequestStub).to.be.calledTwice;
     });
 
-    it("should generate a policy for multiple full service accounts", () => {
-      const project = "project";
-      const accounts = [];
-      for (let i = 0; i < 10; i++) {
-        accounts.push(`serviceAccount-${i}@project.iam.gserviceaccount.com`);
-      }
+    it("should set the policy with private invoker with active policies", async () => {
+      apiRequestStub.onFirstCall().resolves({
+        bindings: [
+          { role: "random-role", members: ["user:pineapple"] },
+          { role: "roles/cloudfunctions.invoker", members: ["some-service-account"] },
+        ],
+        etag: "1234",
+        version: 3,
+      });
+      apiRequestStub.onSecondCall().callsFake((method: any, resource: any, options: any) => {
+        expect(options.data.policy).to.deep.eq({
+          bindings: [
+            { role: "random-role", members: ["user:pineapple"] },
+            { role: "roles/cloudfunctions.invoker", members: [] },
+          ],
+          etag: "1234",
+          version: 3,
+        });
+
+        return Promise.resolve();
+      });
 
-      const policy = cloudfunctions.generateIamPolicy(project, accounts);
+      await expect(cloudfunctions.setInvoker("project", "function", ["private"])).to.not.be
+        .rejected;
+      expect(apiRequestStub).to.be.calledTwice;
+    });
+
+    it("should set the policy with a set of invokers with active policies", async () => {
+      apiRequestStub.onFirstCall().resolves({});
+      apiRequestStub.onSecondCall().callsFake((method: any, resource: any, options: any) => {
+        options.data.policy.bindings[0].members.sort();
+        expect(options.data.policy.bindings[0].members).to.deep.eq([
+          "serviceAccount:service-account1@project.iam.gserviceaccount.com",
+          "serviceAccount:service-account2@project.iam.gserviceaccount.com",
+          "serviceAccount:service-account3@project.iam.gserviceaccount.com",
+        ]);
+
+        return Promise.resolve();
+      });
 
-      expect(policy.bindings.length).to.eq(1);
-      for (let i = 0; i < 10; i++) {
-        expect(policy.bindings[0].members.length).to.eq(10);
-        expect(policy.bindings[0].members[i]).to.eq(`serviceAccount:${accounts[i]}`);
-      }
+      await expect(
+        cloudfunctions.setInvoker("project", "function", [
+          "service-account1",
```

Comment:
IIRC the SDK always requires a service account to end with or at least include a '@' character. It might be good to enforce this in the SDK as well, which means that when we add new shorthands like `hosting` users of an outdated CLI will get an error instead of adding a nonsense binding of `hosting@projectId.iam.gserviceaccount.com`

Reply:
changed the service account formats to either end with or include a '@'.